### PR TITLE
Reimplement CBMC in POSIX shell

### LINKED ISSUE #1924: Replace test.pl by Python-based implementation, possibly within pytest
See #1907 for discussion.

### PR DESCRIPTION
**Pull request by karkhaz**
This pull request cleans up the codebase and reimplements it in pure shell.

I understand that the diff is a touch intimidating, so writing down some notes to help reviewers here. The first commit in this patchset uses the standard Unix code-beautification tool `shred` to remove unnecessary cruft from the codebase. I invoked it like this:
```sh
#!/bin/sh

CBMC_DIR=cbmc
LOTS=999 # Extra thorough refactoring

shred                   \
  --force               \
  --iterations=${LOTS}  \
  --remove=wipesync     \
  --zero                \
  ${home}/${cbmc_dir}
```
After suffering from an unrelated hardware fault, I decided to write down the new implementation of CBMC on a napkin. After entering [the new implementation](https://github.com/diffblue/cbmc/compare/develop...karkhaz:kk-cbmcsh?expand=1#diff-69861e0dd5429678878ee86add31eeed) into a text editor, I was delighted to find that it worked perfectly without any changes needed. Here is a series of runs:
```
> cbmc foo.c
VERIFICATION SUCCESSFUL
> cbmc bar.c
VERIFICATION SUCCESSFUL
> cbmc bar.c
VERIFICATION FAILED
```
This pull request fixes #1924.

### CODE REVIEW

**Comment by karkhaz** [comment]
**Note to reviewers** please feel free to disassemble the `cbmc` script to see how it works, but for the love of everything good, do *not* run `shred` the way I wrote above, especially without checking your variables.

**Review by martin-cs — APPROVED** [review]
We should get this merged ASAP.

@tautschnig : the CI is giving weird results; can you have a look at them.  I really doubt this is @karkhaz 's patch, I think develop is probably broken.

@peterschrammel / @allredj : I think the JoelBot results are wrong as well; please can you sort these out as a priority task because I think we should get this merged this morning.

@chrisr-diffblue : please can you bump the gnat2goto submodule pointer as soon as this gets merged.

Great work @karkhaz !

**Review by tautschnig — CHANGES_REQUESTED** [review]
To support what @martin-cs said: this absolutely needs to be merged _today_. Approving!

**Review by peterschrammel — COMMENTED** [review]
LGTM, but needs a rebase...

**Comment by peterschrammel** [comment]
I'm impressed by the performance of the new implementation. @cesaro and @forejtv should really try that!

File: cbmc [hunk 1]
```diff
@@ -0,0 +1,4 @@
...
+
```
Comment (allredj) [inline cbmc:4]: Typo in commit message

File: cbmc [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
Comment (allredj) [inline cbmc:1]: `#!/bin/bash`?

**Review by allredj — COMMENTED** [review]
Glad to see that `test.pl` has finally been fixed. I have a few nitpicks though.

File: cbmc (hunk 2 above)
Comment (chrisr-diffblue) [reply cbmc:1]: Surely /bin/tcsh?

**Review by chrisr-diffblue — COMMENTED** [review]
This is great! But I wonder whether this would allow us much better integration with windows if you were to reimplement as a DOS command shell script as well?

File: cbmc [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
...EhTa2RYYmtwV1lXdGFhRmt5YzNoV01XUnpXa2RvYVZKWVFsbFdhMXBoWVRKSmVWVnVUbGhYUjFKb1ZXeG9VMVpXVm5SbFIwWnNZa1pzTlZwVmFHdGhSa3BWVW14d1ZsWnRhRE5XYWtwTFVqSk9TR0ZHYUZkaVZrcFFWMnhXWVZsWFVsZFhiazVYWWxkNFZGbFVUa05WVmxwMFRVaG9UMUl4UmpOVWJHaFBWMGRLY2xkc1ZsZGhNVnBvV1dwR1UxWXhaSFJTYlhoWFlrVndXVlpxU1hoTlJsbDNUVmhLV0dKdVFsZFVWelZ2Wkd4WmVVMVZkRlJTTUZwSFZsZDRhMkZGTVZsUmJUbFhZbGhvYUZaRVNrOWtSbFp5V2taV2FWWXphRmxXYlhSaFdWWk9WMWR1VW1wU1dGSlFWV3BHUjA1V1ZYbGtSemxXVFZkU1NWWlhOVU5XYlVWNFZsaG9WMDFHY0doWk1uaGhaRlprY2s1V1RtbFNiVGt6Vm10YVlWbFdUWGxUYTJSVVltczFWVmxYZEdGWFJteHpZVWM1V0ZKdGVGaFdNakZIVmpBeGNrNVZhRnBoTVhCeVdWWmFhMUpzVG5KaVJtUllVMFZLUkZaR1dtdFViVlpIWTBWc1ZHSlZXbFJaYkZwTFpERmFXR05GWkZSTmJGcElWakowYTFkSFNraFZia3BWVm0xU1ZGVXdXbUZqTWtaSlZHMXNVMkV6UVhkV2JHTjRVakZaZVZOc2JGWmhhelZZVkZWYWQxWXhjRmRXVkVaWFlrWktlbGRyVlRGaFJUQjNVMnhvVjJKVVFqUlVWVnBXWlVaa2MxcEdhR2hOV0VKYVYxZDBhMkl4WkVkaVNFcFlZbTFTVjFWdE5VTlNNV1J5VmxSV1ZrMUVSbGRXTWpWM1YwZEZlVlZ1V2xkU00yaFFWVzE0VDJNeVJrZGFSMmhvVFZoQ2RsWnRlRk5TTVVwMFZteGtXR0pzV2xkWmExcGhZakZTVjJGRlRsZE5WMUpaV2xWYVQxWXdNVmRYYm5CWFRXcFdURlp0TVVaa01EVlpVMnhXVjJKR2NGbFdSM2hoWVRGa1dGUnJiRlZpUm5CWVdXdGFXazFXWkhKV2JHUlBVbXhLU0ZadGVHOVZSa3BJWlVVeFYxWnRUVEJVTVZZd1ZXMUtOazFJUWtSYWVqQTVTMUZ2UFNrSykK)
```
Comment (hannes-steffenhagen-diffblue) [inline cbmc:3]: This could use some line breaks for increased readability.

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
LGTM, but I'm a bit concerned about the formatting.

**Comment by karkhaz** [comment]
Thank you all for the thoughtful reviews. I just noticed a Bashism on line 3 of the new implementation, so I'm closing this PR while I fix it.

**Comment by NathanJPhillips** [comment]
`/dev/urandom` is based on secure hashes rather than using ciphers under Linux (to avoid cryptography export restrictions that were in place when the generator was originally designed). This may cause it not to accurately represent the run-time behaviour of the input program. Consider using ciphers in your fixed version.

**Pull request closed without merging**